### PR TITLE
Fix issue 18995 - Make sure GC calls destructor when using std.array.array

### PR DESCRIPTION
**Pull request by schveiguy**
Note that I'm calling the delete operator indirectly here, since `delete` has been deprecated. `GC.free` does not run finalizers on the memory. I'm not sure exactly what `GC.runFinalizers` does (it seems to have to do with shared libraries), so I didn't use that.

### CODE REVIEW

**Comment by ZombineDev** [comment]
I think it's better to leave the call to `uninitializedArray` in `array` as it is and replace the call to `GC.malloc` in `arrayAllocImpl` (that's what `uninitializedArray` and `minimallyInitializedArray` use)
to `_d_newarrayU` (from `rt.lifetime`). After all druntime provides the correct implementation for `uninitializedArray`, so I see no reason not to use it.

File: std/array.d [hunk 1]
```diff
@@ -219,6 +223,34 @@ if (isPointer!Range && isIterable!(PointerTarget!Range) && !isNarrowString!Range
...
+
+    import std.algorithm.iteration : map;
+    import std.range : iota;
+
+    auto arr = iota(3).map!(a => S(a)).array;
+    assert(nAlive == 3);
+
+    // No good way to ensure the GC frees this, just call the lifetime function
+    // directly. If delete wasn't deprecated, this is what delete would do.
+    _d_delarray_t(cast(void[]*)&arr, typeid(S));
```
[the remark this answers is not in the document]
Comment (ZombineDev) [reply std/array.d]: The problem is that `destroy` works too good as it is a template and calls the destructors directly. The issue is that unless the GC knows that the memory allocation is used for an array of objects with destructors it won't call those destructors during collection.

[the remark this answers is not in the document]
Comment (ZombineDev) [reply std/array.d]: @schveiguy have you tried calling `GC.collect` directly? I have seen unittests in dmd and druntime that force a collection to ensure destructors are called.

**Comment by schveiguy** [comment]
> to _d_newarrayU (from rt.lifetime).

Ah, I didn't know this existed! However, here's one concern I have -- if the type has indirections, an uninitialized block will still be scanned (the part that isn't used). But in the interest of keeping this PR minimal, I'll just switch the internal function, and file a bug on that. It took me a long time to find out why there are two *identical* `uninitializedArray` instances -- one for items with pointers, and one for items without. The difference is one is `@trusted`, and one is `@system`. IMO, the one that has indirections should still initialize the portion not being allocated.

File: std/array.d (hunk 1 above)
[the remark this answers is not in the document]
Comment (schveiguy) [reply std/array.d]: I haven't, but this seems inferior to me. If I use `GC.collect`, there is a chance (small chance) that it won't actually be collected. These kinds of calls result in at some point code that tested fine seemingly breaks, maybe due to randomness, maybe due to changes in environment. The result is that PRs fail in confusing ways. I'd much rather call the function directly.

I did run the new unittest without my changes to ensure that it fails.

**Comment by schveiguy** [comment]
@ZombineDev updated to fix `arrayAllocImpl` instead, I found a curious issue with an old test, just commented it out, as I think the test was invalid.

File: std/array.d (hunk 1 above)
[the remark this answers is not in the document]
Comment (wilzbach) [reply std/array.d]: > If delete wasn't deprecated, this is what delete would do.

Why not call `__delete`?
https://dlang.org/phobos/core_memory.html#.__delete

[the remark this answers is not in the document]
Comment (schveiguy) [reply std/array.d]: > Why not call `__delete`?

No, that's the same as calling destroy. I want to do what the GC would do, without having any static introspection.

**Comment by ZombineDev** [comment]
Thanks @schveiguy and sorry for the delayed merging. I have been contemplating on the original reason why the test with an assert in the destructor was added, but it looks like it's simply straight copy-paste from a reduced bug test case and not something that was supposed be an actual phobos unit test.

**Pull request merged**